src/mgmt/srv.c

```diff
@@ -72,314 +72,430 @@ static struct daos_rpc_handler mgmt_handlers[] = {
+	uint8_t	*body;
+	size_t	len;
 
-	/* Response status is populated with SUCCESS on init. */
-	mgmt__daos_resp__init(daos_resp);
+	len = mgmt__daos_resp__get_packed_size(daos_resp);
+	D_ALLOC(body, len);
+	if (body == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to allocate drpc response body\n");
+		return;
+	}
 
-	/* Unpack the daos request from the drpc call body */
-	pb_rank = mgmt__daos_rank__unpack(
-		NULL, drpc_req->body.len, drpc_req->body.data);
+	if (mgmt__daos_resp__pack(daos_resp, body) != len) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Unexpected num bytes for daos resp\n");
+		return;
+	}
 
-	if (pb_rank == NULL) {
-		daos_resp->status = MGMT__DAOS_REQUEST_STATUS__ERR_UNKNOWN;
-		D_ERROR("Failed to extract rank from request\n");
+	/* Populate drpc response body with packed daos response */
+	drpc_resp->body.len = len;
+	drpc_resp->body.data = body;
+}
 
+static void
+process_killrank_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
+{
+	Mgmt__DaosRank		*req = NULL;
+	Mgmt__DaosResp		*resp = NULL;
+
+	/* Unpack the inner request from the drpc call body */
+	req = mgmt__daos_rank__unpack(
+		NULL, drpc_req->body.len, drpc_req->body.data);
+	if (req == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to unpack req (kill rank)\n");
 		return;
 	}
 
 	D_DEBUG(DB_MGMT, "Received request to kill rank (%u) on pool (%s)\n",
-		pb_rank->rank, pb_rank->pool_uuid);
+		req->rank, req->pool_uuid);
+
+	D_ALLOC_PTR(resp);
+	if (resp == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to allocate daos response ref\n");
+		mgmt__daos_rank__free_unpacked(req, NULL);
+		return;
+	}
+
+	/* Response status is populated with SUCCESS on init. */
+	mgmt__daos_resp__init(resp);
 
 	/* TODO: do something with request and populate daos response status */
 
-	mgmt__daos_rank__free_unpacked(pb_rank, NULL);
+	mgmt__daos_rank__free_unpacked(req, NULL);
+	pack_daos_response(resp, drpc_resp);
+	D_FREE(resp);
 }
 
 static void
-process_setrank_request(Drpc__Call *drpc_req, Mgmt__DaosResp *daos_resp)
+process_setrank_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 {
-	Mgmt__SetRankReq	*daos_req = NULL;
+	Mgmt__SetRankReq	*req = NULL;
+	Mgmt__DaosResp		*resp = NULL;
 	int			rc;
 
-	/* Response status is populated with SUCCESS on init. */
-	mgmt__daos_resp__init(daos_resp);
-
-	/* Unpack the daos request from the drpc call body */
-	daos_req = mgmt__set_rank_req__unpack(
+	/* Unpack the inner request from the drpc call body */
+	req = mgmt__set_rank_req__unpack(
 		NULL, drpc_req->body.len, drpc_req->body.data);
+	if (req == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to unpack req (set rank)\n");
+		return;
+	}
 
-	if (daos_req == NULL) {
-		daos_resp->status = MGMT__DAOS_REQUEST_STATUS__ERR_UNKNOWN;
-		D_ERROR("Failed to extract request\n");
+	D_DEBUG(DB_MGMT, "Received request to set rank to %u\n",
```

Comment:
I wonder if it would be worth logging these as D_INFO - since these are administrative actions that might be interesting to audit even if someone isn't debugging.

Reply:
https://github.com/daos-stack/daos/pull/816